### ISSUE TITLE
WPT server is slow

### ISSUE THREAD

**Issue by youennf** [w3c/web-platform-tests#8391]
Following on https://bugs.webkit.org/show_bug.cgi?id=178277 and https://lists.webkit.org/pipermail/webkit-dev/2017-November/029768.html, the WPT server is noticeably slower than Apache, something like 3 to 10 times slower than Apache for a single file download.
It also seems that serving WPT tests through Apache may be 25% or more faster than using WPT server (for those tests not relying on WPT server nice features).

Two issues have been identified:
- Apache allows resource caching. WPT server started to do so for testharness.js recently.
- Apache provides Content-Length header, not WPT server (https://github.com/w3c/wptserve/pull/86)
Maybe there are other actions that could be done to help on this.

**Comment by foolip** [comment]
Is the appropriate benchmark to focus on here the time to run the entire test suite in WebKit?

Do you have any numbers for the difference in serving single small and large files that aren't cached? That should tell us something about where to look, i.e. is it a constant overhead to handle a request, inefficient I/O, or something else...

**Comment by jgraham** [comment]
I can ressurrect the wptserve patch to apply the content-length in more cases; I hadn't seen it before but it looks entirely reasonable. But I will note that the last patch I made (to cache `testharness.js`) had no observable impact on the end-to-end time on Gecko infrastructure. So I can only implement things  blind here; I can't tell what will have an impact on WebKit infrastructure. Progress is likely to be much faster if someone from the WebKit project steps up to do the work here.

All of that said, I don't think that "serving a file is as fast as Apache" is a plausible goal. So I'd like to know what the end to end times are and what's would be considered acceptable.

**Comment by jgraham** [comment]
FTR with /FileAPI/url/url_xmlhttprequest.html I can't reproduce a timing difference between Apache and wptserve. I might be doing something wrong ofc.

**Comment by jgraham** [comment]
I applied the `Content-Length` patch to https://treeherder.mozilla.org/#/jobs?repo=try&revision=3616d4e0aa004d8d472fa6b0b68804bd21d9102c Unsurprisingly it breaks some tests (and fixes some too, it  seems), but I can't see any discernable performance difference in a sample of Windows jobs that pass before and after. I would expect Windows to have the worst IO speed and so be most affected by the server performance.

**Comment by youennf** [comment]
@jgraham,  can you point me to the Content-Length patch you used.
I can give it a try on my WebKit setup (using run-webkit-tests) to see the impact.
I think this patch will be an improvement anyway, for instance when running some fetch tests doing 20 redirections through HTTPS on w3c-test.org server.

I also tried the testharness.js patch locally and did not observe any noticeable difference on the IDBIndex limited test set I run. I can try to add more caching locally, like testharnessreport.js and see whether that helps.

I tried running the IDB sub test suite with WPT and Apache and forcing parallelizing as much as possible. In such a case, the time spent is very similar whether served through Apache or WPT server.

**Comment by jgraham** [comment]
The try push above has the patch (applied to mozilla-central) https://hg.mozilla.org/try/rev/966a2c27dccba8b4f37a7b8257d94ddf41365b14. You can grab the diff and edit the paths as required.

**Comment by youennf** [comment]
Just tried the patch on the same IDB test suite.
Tests now run as fast as Apache.

**Comment by youennf** [comment]
> Is the appropriate benchmark to focus on here the time to run the entire test suite in WebKit?

Right, I don't think the goal is to have WPT server be as fast as Apache.
The goal is more that the entire test suite should run at about the same time.
A 25% difference seems worth investigating and fixing if not too complex.

**Comment by foolip** [comment]
I agree, if it's really 25% then that could be very low hanging fruit. Do you have a set of tests that you've done this experiment with that we might try in Chromium as well? If we see similar differences, then maybe we should prioritize this on our end as well.

**Comment by youennf** [comment]
I tried the following test suite: https://bug-178277-attachments.webkit.org/attachment.cgi?id=327431 
I only checked that test results were the same when served by Apache and WPT.

I would think ref tests should also be good candidates to show any potential difference since they might take less time to execute.

**Comment by Hexcles** [comment]
On #testing IRC today a few other potential performance issues were discussed:

* IPC is slower on XNU than on Linux, which wptserve uses quite a lot.
* Exhaustion of ephemeral ports (as described in this [article](http://danielmendel.github.io/blog/2013/04/07/benchmarkers-beware-the-ephemeral-port-limit/)), and perhaps max open files. @youennf , @gsnedders said you ran into this issue before, can you provide more insights?

And they are by no means exhaustive...

**Comment by gsnedders** [comment]
> Exhaustion of ephemeral ports (as described in this article), and perhaps max open files. @youennf , @gsnedders said you ran into this issue before, can you provide more insights?

that's https://github.com/w3c/web-platform-tests/issues/8369

**Comment by rwaldron** [comment]
cc @mattl

**Comment by foolip** [comment]
I added the roadmap label to this, not because we should work on it this quarter, but because we should at least investigate in Q1 how slow it is and if there are any easy improvements to be made. I've dropped a link in [Ecosystem Infra 2018 Q1 OKRs](https://docs.google.com/document/d/1ojlLDiRruI4e95Z7JQ10JQPRMiQHd_40rBkYCL6IJqw/edit?usp=sharing) for now.

**Comment by foolip** [comment]
@Hexcles, do you have an update on "Measure overhead of wpt serve in Chromium and WebKit"? I believe work on this hasn't been started yet, but this came up in issue triage today.

**Comment by Hexcles** [comment]
@foolip You're right. I plan to do the benchmark in March.

**Comment by jgraham** [comment]
March happened, is there any update here?

**Comment by Hexcles** [comment]
Oops, unfortunately no... This has been deprioritized since our strategic planning in March regarding the running infra.We'll be focusing on wpt.fyi and its supporting APIs for now.

(But I do wish to get back to this at some point.)
